pkg/vmprovider/providers/vsphere2/virtualmachine/configspec.go

```diff
@@ -106,6 +114,48 @@ func CreateConfigSpec(
 	return &configSpec
 }
 
+func determineHardwareVersion(
+	vm *vmopv1.VirtualMachine,
+	configSpec *types.VirtualMachineConfigSpec,
+	vmImageStatus *vmopv1.VirtualMachineImageStatus) int32 {
+
+	vmMinVersion := vm.Spec.MinHardwareVersion
+
+	var configSpecVersion int32
+	if configSpec.Version != "" {
+		configSpecVersion = util.ParseVirtualHardwareVersion(configSpec.Version)
+	}
+
+	if configSpecVersion == 0 {
+		// A VM Class with an embedded ConfigSpec should have the version set, so this is
+		// a ConfigSpec we created from the HW devices in the class. If the image's version
+		// is too old to support passthrough devices or PVCs if configured, bump the version
+		// so those devices will work.
+		var imageVersion int32
+		if vmImageStatus != nil && vmImageStatus.HardwareVersion != nil {
+			imageVersion = *vmImageStatus.HardwareVersion
+		}
+
+		var minVerFromDevs int32
+		if util.HasVirtualPCIPassthroughDeviceChange(configSpec.DeviceChange) {
+			minVerFromDevs = max(imageVersion, constants.MinSupportedHWVersionForPCIPassthruDevices)
+		} else if hasPVC(vm) {
+			// This only catches volumes set at VM create time.
+			minVerFromDevs = max(imageVersion, constants.MinSupportedHWVersionForPVC)
+		}
+
+		// If both are zero, VC will use the cluster's default version.
+		return max(vmMinVersion, minVerFromDevs)
```

Comment:
Should we not just be setting `configSpecVersion = minVerFromDevs` and letting it fall through outside the `if` block?

Reply:
I've gone back & forth a bit on how to organize this function. I think this early return works out best b/c a `configSpec.Version == ""` is kind of the exceptional case nowadays, and to fall through here  we need another `configSpec.Version != ""` check to preserve the prior & desired behavior of when to update that field.

Reply:
Sure.